### PR TITLE
[ci] Add local dev flags to e2e script

### PR DESCRIPTION
**Pull request by ravisantoshgudimetla**
As of now, after the test suite has been run, the
Windows nodes would be deleted as well. This commit
ensures that
- We have a flag which retains the Windows node
 created as part of test suite. Please note that
 we're skipping the part where the delete
 suite is being called in case this flag has been
 set.
- We have flag to specify the number of number of
  Windows nodes to be created as part of test run
- We won't download operator-sdk if it's already
  available in the PATH

Example command to spin up 2 Windows nodes and retain them after test run:

```
hack/run-ci-e2e-test.sh -s -k "openshift-dev" -n 2"
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -42,9 +39,14 @@ for our CI purposes.
...
 ```shell script
-hack/run-ci-e2e-test.sh
+hack/run-ci-e2e-test.sh -k "openshift-dev"
 ```
-
+Additional flags that can be passed to `hack/run-ci-e2e-test.sh` are
+- `-s` to skip the deletion of Windows nodes that are created as part of test suite run
+- `-n` to represent the number of Windows nodes to be created for test run
+- `-k` to represent the key pair that will be used during e2e run. This private key path in this pair should match with 
+       what we have in `KUBE_SSH_KEY_PATH` The default value points to `libra` which we use in our CI
```
Comment (aravindhp) [inline README.md:48]: This is confusing. From the example above it seems you are asking for the key pair name in the cloud account but you are not being clear about the connection with `KUBE_SSH_KEY_PATH`.

Please add the example you have included in the commit message here too.

File: hack/run-ci-e2e-test.sh [hunk 2]
```diff
@@ -3,21 +3,54 @@ set -o errexit
...
 
 export CGO_ENABLED=0
 export CLUSTER_ADDR=$(oc cluster-info | head -n1 | sed 's/.*\/\/api.//g'| sed 's/:.*//g')
 
+while getopts ":n:k:s" opt; do
+  case ${opt} in
+    n ) # process option for providing existing VM credentials
+      NODE_COUNT=$OPTARG
+      ;;
+    k )
```
Comment (aravindhp) [inline hack/run-ci-e2e-test.sh:21]: Missing doc

File: README.md (hunk 1 above)
Comment (ravisantoshgudimetla) [reply README.md:48]: `This private key path in this pair should match with what we have in `KUBE_SSH_KEY_PATH``. You mean I should elaborate on that?

File: hack/run-ci-e2e-test.sh [hunk 3]
```diff
@@ -3,21 +3,54 @@ set -o errexit
 set -o nounset
 set -o pipefail
 
+# TODO: Add input validation
```
[the remark this answers is not in the document]
Comment (ravisantoshgudimetla) [reply hack/run-ci-e2e-test.sh:6]: I think it's ok even if we don't the validation as long as the documentation is clear. I'll remove the TODO

File: hack/run-ci-e2e-test.sh (hunk 2 above)
Comment (ravisantoshgudimetla) [reply hack/run-ci-e2e-test.sh:21]: Will add it

File: README.md (hunk 1 above)
Comment (aravindhp) [reply README.md:48]: What does `This private key path in this pair` even mean? You should make it less confusing. Reading the explanation it seemed like `-k` can be used to specify the key itself. You need to call out that this specific to key-pair names in the AWS account and that it should map to the private key specified in `KUBE_SSH_KEY_PATH`.

File: hack/run-ci-e2e-test.sh [hunk 4]
```diff
@@ -3,21 +3,54 @@ set -o errexit
...
+done
+
 # Copy the cloud credentials and KUBESSH key path so that they can be used by operator
 cp $AWS_SHARED_CREDENTIALS_FILE /etc/cloud/credentials
 cp $KUBE_SSH_KEY_PATH /etc/private-key/private-key.pem
 
 # Get operator-sdk binary.
-# TODO: Make this to download the same version we have in go dependencies in gomod
-wget -O /tmp/operator-sdk https://github.com/operator-framework/operator-sdk/releases/download/v0.15.2/operator-sdk-v0.15.2-x86_64-linux-gnu && chmod +x /tmp/operator-sdk
+# We don't care about the operator-sdk version we're using if running locally.
```
Comment (mansikulkarni96) [inline hack/run-ci-e2e-test.sh:39]: Could be reframed, a little confusing as of now something like, we can use any operator-sdk version if we are using it for running locally.

File: hack/run-ci-e2e-test.sh [hunk 5]
```diff
@@ -3,21 +3,54 @@ set -o errexit
...
+
+# Set default value for node count
+NODE_COUNT=${NODE_COUNT:-1}
+
 cd $WMCO_ROOT
-oc create -f deploy/namespace.yaml 
-/tmp/operator-sdk test local ./test/e2e --up-local --namespace=windows-machine-config-operator --go-test-flags "-v -timeout 60m"
+oc create -f deploy/namespace.yaml
+# The bool flags in golang does not respect key value pattern. They follow key=value pattern.
+# -flag x  // non-boolean flags only(https://golang.org/pkg/flag/)
```
Comment (mansikulkarni96) [inline hack/run-ci-e2e-test.sh:52]: this doc is a little confusing.

File: test/e2e/main_test.go [hunk 6]
```diff
@@ -27,6 +32,10 @@ var (
 type globalContext struct {
 	// numberOfNodes to be used for the test suite.
 	numberOfNodes int
+	// skipNodeDeletion flag allows skipping deletion of nodes within the test suite
```
Comment (mansikulkarni96) [inline test/e2e/main_test.go:35]: should this doc be consistent with above doc for skipNodeDeletion?

Comment (ravisantoshgudimetla) [reply test/e2e/main_test.go:35]: no preference but I can change it.

File: hack/run-ci-e2e-test.sh [hunk 7]
```diff
@@ -3,21 +3,54 @@ set -o errexit
...
 cp $KUBE_SSH_KEY_PATH /etc/private-key/private-key.pem
 
 # Get operator-sdk binary.
-# TODO: Make this to download the same version we have in go dependencies in gomod
-wget -O /tmp/operator-sdk https://github.com/operator-framework/operator-sdk/releases/download/v0.15.2/operator-sdk-v0.15.2-x86_64-linux-gnu && chmod +x /tmp/operator-sdk
+# We don't care about the operator-sdk version we're using if running locally.
+if ! type operator-sdk > /dev/null; then
+  # TODO: Make this to download the same version we have in go dependencies in gomod
+  wget -O /tmp/operator-sdk https://github.com/operator-framework/operator-sdk/releases/download/v0.15.2/operator-sdk-v0.15.2-x86_64-linux-gnu && chmod +x /tmp/operator-sdk
+  export PATH=$PATH:/tmp
```
Comment (sebsoto) [inline hack/run-ci-e2e-test.sh:43]: I'm not sure if this should be done, it could cause other side effects. Might be better to create an alias or something?

Comment (ravisantoshgudimetla) [reply hack/run-ci-e2e-test.sh:43]: I thought about this but I looked at the pod in CI which is controlled environment and it shouldn't cause an issue and in our machine when we are running it might be ok considering the `PATH` env variable is limited to scope of this bash script. However if you want this changed I can do it.

File: hack/run-ci-e2e-test.sh (hunk 4 above)
Comment (ravisantoshgudimetla) [reply hack/run-ci-e2e-test.sh:39]: I mean we're not validating if the `operator-sdk` which developer has in his/her machine matches with 0.15.2 which is what we have tested for.

File: test/e2e/main_test.go (hunk 6 above)
Comment (mansikulkarni96) [reply test/e2e/main_test.go:35]: This would be just a nit.

File: hack/run-ci-e2e-test.sh (hunk 5 above)
Comment (ravisantoshgudimetla) [reply hack/run-ci-e2e-test.sh:52]: Which part?

File: README.md [hunk 8]
```diff
@@ -42,9 +39,19 @@ for our CI purposes.
 ```
 Once the above variables are set and the /payload directory has been populated, run the following script:
 ```shell script
-hack/run-ci-e2e-test.sh
+hack/run-ci-e2e-test.sh -k "openshift-dev"
```
Comment (aravindhp) [inline README.md:42]: Add a line mentioning that we are assuming developers are using the `openshift-dev` key in their AWS account.

File: hack/run-ci-e2e-test.sh [hunk 9]
```diff
@@ -7,21 +7,61 @@ set -o pipefail
...
 cp $AWS_SHARED_CREDENTIALS_FILE /etc/cloud/credentials
 cp $KUBE_SSH_KEY_PATH /etc/private-key/private-key.pem
 
-# Get operator-sdk binary.
-# TODO: Make this to download the same version we have in go dependencies in gomod
-wget -O /tmp/operator-sdk https://github.com/operator-framework/operator-sdk/releases/download/v0.15.2/operator-sdk-v0.15.2-x86_64-linux-gnu && chmod +x /tmp/operator-sdk
+# Download the operator-sdk binary only if not is not already available
+# We do not validate the version of operator-sdk if it is available already
+if ! type operator-sdk > /dev/null; then
+  # TODO: Make this to download the same version we have in go dependencies in gomod
```
Comment (aravindhp) [inline hack/run-ci-e2e-test.sh:44]: `Make this download...` No need for the `to`.

File: hack/run-ci-e2e-test.sh [hunk 10]
```diff
@@ -7,21 +7,61 @@ set -o pipefail
...
+  wget -O /tmp/operator-sdk https://github.com/operator-framework/operator-sdk/releases/download/v0.15.2/operator-sdk-v0.15.2-x86_64-linux-gnu && chmod +x /tmp/operator-sdk
+  # To expand alias
+  shopt -s expand_aliases
```
Comment (aravindhp) [inline hack/run-ci-e2e-test.sh:47]: Why is this needed?

File: hack/run-ci-e2e-test.sh [hunk 11]
```diff
@@ -7,21 +7,61 @@ set -o pipefail
...
+KEY_PAIR_NAME=${KEY_PAIR_NAME:-"libra"}
+
 cd $WMCO_ROOT
-oc create -f deploy/namespace.yaml 
-/tmp/operator-sdk test local ./test/e2e --up-local --namespace=windows-machine-config-operator --go-test-flags "-v -timeout 60m"
+oc create -f deploy/namespace.yaml
+# The bool flags in golang does not respect key value pattern. They follow key=value pattern.
+# -flag x  // non-boolean flags only(https://golang.org/pkg/flag/)
+operator-sdk test local ./test/e2e --up-local --namespace=windows-machine-config-operator --go-test-flags "-v -timeout=60m -node-count=$NODE_COUNT $SKIP_NODE_DELETION -ssh-key-pair=$KEY_PAIR_NAME"
 oc delete -f deploy/namespace.yaml
```
Comment (PratikMahajan) [inline hack/run-ci-e2e-test.sh:65]: if the test fails, it crashed the script and fails to delete the namespace. We can catch the error and then delete the namespace.

Comment (aravindhp) [reply hack/run-ci-e2e-test.sh:65]: That needs to fixed in a separate PR and not in this one. While fixing that we also need to only create the namespace if it doesn't already exist to help with local development.

File: README.md (hunk 8 above)
Comment (ravisantoshgudimetla) [reply README.md:42]: Added a line a below.

File: hack/run-ci-e2e-test.sh [hunk 12]
```diff
@@ -7,21 +7,61 @@ set -o pipefail
...
+done
+
 # Copy the cloud credentials and KUBESSH key path so that they can be used by operator
 cp $AWS_SHARED_CREDENTIALS_FILE /etc/cloud/credentials
 cp $KUBE_SSH_KEY_PATH /etc/private-key/private-key.pem
 
-# Get operator-sdk binary.
-# TODO: Make this to download the same version we have in go dependencies in gomod
-wget -O /tmp/operator-sdk https://github.com/operator-framework/operator-sdk/releases/download/v0.15.2/operator-sdk-v0.15.2-x86_64-linux-gnu && chmod +x /tmp/operator-sdk
+# Download the operator-sdk binary only if not is not already available
```
[the remark this answers is not in the document]
Comment (aravindhp) [reply hack/run-ci-e2e-test.sh]: You have not addressed this.

File: hack/run-ci-e2e-test.sh [hunk 13]
```diff
@@ -7,21 +7,61 @@ set -o pipefail
...
+NODE_COUNT=${NODE_COUNT:-1}
+SKIP_NODE_DELETION=${SKIP_NODE_DELETION:-"-skip-node-deletion=false"}
+KEY_PAIR_NAME=${KEY_PAIR_NAME:-"libra"}
+
 cd $WMCO_ROOT
-oc create -f deploy/namespace.yaml 
-/tmp/operator-sdk test local ./test/e2e --up-local --namespace=windows-machine-config-operator --go-test-flags "-v -timeout 60m"
+oc create -f deploy/namespace.yaml
+# The bool flags in golang does not respect key value pattern. They follow -flag=x pattern.
+# -flag x is only allowed for non-boolean flags only(https://golang.org/pkg/flag/)
```
Comment (mansikulkarni96) [inline hack/run-ci-e2e-test.sh:63]: extra only in this line.
